### PR TITLE
Use mpirun when testing MPI

### PR DESCRIPTION
**Pull request by aragilar**
This provides more freedom to specify HDF5 locations (which I needed for testing the MPI builds), and runs the tests under MPI (apart from those known to not work, e.g. those using subprocess). I fixed the existing MPI tests, and added some docs about how to test with MPI.

This depends on https://pypi.org/project/pytest-mpi/, which I created because it seems that the `conftest.py` must the at the root level for cli args to work (see https://stackoverflow.com/questions/52447767/using-pytest-addoptions-in-a-non-root-conftest-py), I'm happy to move it under the h5py org and/or give people access (on github, pypi and other services) if people think that's appropriate (I plan on adding docs and tests, happy to restructure it also, I basically used an existing pytest plugin I wrote as a template).

Before this gets merged, can people run this on their systems (and if report any lockups, the tests should run very fast, so if you're waiting a while, there's probably a problem).

#1206 will need to be rebased on this (let me know if there are more features we should expose with pytest-mpi).

### CODE REVIEW

**Comment by takluyver** [comment]
I don't have time to look at this in detail right now, but this sounds excellent - having a decent testing infrastructure for MPI will make things like #1206 much simpler to address.

For the plugin: I think it's fine to add it to the h5py org if you'd like to maintain it collaboratively. But we can also depend on your project if you'd rather keep it separate.

**Comment by aragilar** [comment]
I've switched back to using the system HDF5 (on azure now, rather than travis). I won't do any more experimentation on this branch, I'll leave getting a newly-built version of HDF5 to a different PR/branch.

File: docs/contributing.rst [hunk 1]
```diff
@@ -354,3 +354,44 @@ In the Cython code, these show up as "preprocessor" defines ``MPI`` and
...
+1. Have ``tox`` installed (e.g. via ``pip install tox``)
+2. Have HDF5 built with MPI as per :ref:`build_mpi`
+
+Then running::
+
+   $ CC='mpicc' HDF5_MPI=ON tox -e py37-test-deps-mpi4py
+
+should run the tests. You may need to pass ``HDF5_DIR`` depending on the
+location of the HDF5 with MPI support. You can choose which python version to
+build against by changing py37 (e.g. py27 runs python 2.7, this is a tox
```
Comment (takluyver) [inline docs/contributing.rst]: We might want to use e.g. py36 as the example, since we don't support Python 2.7 any more.

File: docs/contributing.rst [hunk 2]
```diff
@@ -354,3 +354,44 @@ In the Cython code, these show up as "preprocessor" defines ``MPI`` and
...
+If you get an error similar to::
+
+   There are not enough slots available in the system to satisfy the 4 slots
+   that were requested by the application:
+     python
+
+   Either request fewer slots for your application, or make more slots available
+   for use.
+
+then you need to reduce the number of MPI processes you are asking MPI to use.
```
Comment (takluyver) [inline docs/contributing.rst]: How do you control this? What's the default?

File: setup_configure.py [hunk 3]
```diff
@@ -86,12 +89,18 @@ class configure(Command):
 
     user_options = [('hdf5=', 'h', 'Custom path to HDF5'),
                     ('hdf5-version=', '5', 'HDF5 version "X.Y.Z"'),
+                    ('hdf5-includedir=', 'i', 'path to HDF5 headers'),
+                    ('hdf5-libdir=', 'l', 'path to HDF5 library'),
+                    ('hdf5-pkgconfig-name=', 'p', 'name of HDF5 pkgconfig file'),
```
Comment (takluyver) [inline setup_configure.py]: It looks like the added config options are no longer required for the MPI testing. Can I suggest that we pull these additions out as a separate PR, and do the minimum necessary in this one to effectively test with MPI?

File: h5py/tests/test_filters.py [hunk 4]
```diff
@@ -61,6 +62,7 @@ def test_wr_scaleoffset_fletcher32(self):
                                   )
 
 
+@pytest.mark.mpi_break
```
Comment (takluyver) [inline h5py/tests/test_filters.py]: For clarity, I might suggest calling the marker `mpi_skip` or `mpi_xfail` - you might even allow both to distinguish "this shouldn't work" vs "this should work but doesn't".

File: docs/contributing.rst (hunk 2 above)
Comment (takluyver) [reply docs/contributing.rst]: I see in tox.ini that the default is 2. I imagine reducing this doesn't make much sense - testing with 1 process is going to miss things. Maybe we should briefly mention how to make more slots available.

Comment (RizzerOnGitHub) [reply docs/contributing.rst]: The OpenMPI flag 

    export OMPI_MCA_rmaps_base_oversubscribe=1

or `mpirun -oversubscribe` might be useful in this context.

File: docs/contributing.rst (hunk 1 above)
Comment (aragilar) [reply docs/contributing.rst]: Yep, that's probably wise.

File: docs/contributing.rst (hunk 2 above)
Comment (aragilar) [reply docs/contributing.rst]: I've added a section on this mentioning the OMPI env var, I think only OpenMPI has this issue (MPICH apparently doesn't, I've got no idea how to setup MS-MPI and I'm not sure there are any other maintained MPI systems).

File: setup_configure.py (hunk 3 above)
Comment (aragilar) [reply setup_configure.py]: Done, now in #1414.

File: h5py/tests/test_filters.py (hunk 4 above)
Comment (aragilar) [reply h5py/tests/test_filters.py]: I've release a new version of pytest-mpi which adds `mpi_skip` and `mpi_fail`, which should be more understandable.

**Comment by takluyver** [comment]
It looks like the tests crashed immediately while trying to import h5py, but then sat until they timed out after an hour. Before making the tests work, can we ensure that it stops correctly on failure?

**Comment by aragilar** [comment]
I've dropped the CI changes from this, so hopefully the tests now pass (leaving the current MPI tests on travis). I'll look at expanding the CI factors the MPI tests run on in a different PR.

**Comment by takluyver** [comment]
The test changes here look good to me; thanks.
